File: src/survAUC_Hung_Chiang.c
/*
 *  survAUC_Hung_Chiang.c
 *  Daim
 *
 *  Created by Sergej Potapov on 10.10.10.
 *  Copyright 2010 __IMBE__. All rights reserved.
 *  2022-05-18. Updated by F. Bertrand <frederic.bertrand@utt.fr>
 *
 */



#include "utils.h"




void C_Hung_Chiang(double *time, int *n_time, double *stime, double *event, int *n_stime,
				 double *stime_new, double *event_new, int *n_stime_new,
				 double *lpnew, int *n_lpnew, double *ans, double *i_auc)
{
	int i, j, k;

	double *SX, *SXa, *St, *Sta, *Sc, *Sca; 
	double *tmp_status, *Sc_stime, *Sc_event;
	double tmp_nen = 0.0;
	Sc_stime = R_Calloc(*n_stime, double);
	Sc_event = R_Calloc(*n_stime, double);
	SX = R_Calloc(*n_stime, double);
	St = R_Calloc(*n_stime, double);
	Sc = R_Calloc(*n_stime, double);
	tmp_status = R_Calloc(*n_stime, double);
	SXa = R_Calloc(*n_time, double);
	Sta = R_Calloc(*n_time, double);
	Sca = R_Calloc(*n_stime_new, double);
	
	for(i=0; i<*n_stime; i++){
		tmp_status[i] = 1.0;
		Sc_stime[i] = stime[i];
		Sc_event[i] = 1.0 - event[i];
	}
	
	C_km_Daim(St, stime, event, n_stime);
	step_eval2(Sta, time, St, stime, *n_time, *n_stime);
	
	C_km_Daim(SX, stime, tmp_status, n_stime);
	step_eval2(SXa, time, SX, stime, *n_time, *n_stime);
	
	C_km_Daim(Sc, Sc_stime, Sc_event, n_stime);
	step_eval2(Sca, stime_new, Sc, Sc_stime, *n_stime_new, *n_stime);

	/* Calculation of AUC */
	for(k=0; k<*n_time; k++){
		for(i=0; i<*n_lpnew; i++){
			for(j=0; j<*n_lpnew; j++){
				if(i != j && ((event_new[i] && (lpnew[i] > lpnew[j]) && (stime_new[i] <= time[k] && stime_new[j] > time[k])) && (Sca[i] > FLT_EPSILON))){
					ans[k] += 1.0 / (Sca[i]);
				}
			}
		}
		tmp_nen = SXa[k]*(1.0-Sta[k])*(*n_lpnew)*(*n_lpnew-1);
		if(tmp_nen > FLT_EPSILON)
			ans[k] /= tmp_nen;
		else
			ans[k] = 0.0;
	}
	R_Free(SX);R_Free(SXa);R_Free(Sc);R_Free(Sca);R_Free(St);
	R_Free(Sta);R_Free(Sc_event);R_Free(Sc_stime);R_Free(tmp_status);
	/* Calculation of iAUC */
	
	double *f, *S, *S_new;
	f = R_Calloc(*n_time, double);
	S_new = R_Calloc(*n_stime_new, double);
	S = R_Calloc(*n_time, double);
	
	C_km_Daim(S_new, stime_new, event_new, n_stime_new);
	step_eval2(S, time, S_new, stime_new, *n_time,  *n_stime_new);
	
	f[0] = 1.0 - S[0];
	for(i=1; i<*n_time; i++){
		f[i] = S[i-1] - S[i];
	}
	double wT = 0.0;
	for(i=0; i < *n_time; i++){
		if(f[i] > FLT_EPSILON){
			wT += f[i];
		}
	}
	for(i=0; i < *n_time; i++){
		if(wT != 0.0){
			/* cumulative case*/
			if(f[i] > FLT_EPSILON && R_finite(ans[i]) ){
				*i_auc += ans[i] * (f[i]) / wT;
			}
		}
	}
	R_Free(f);R_Free(S);R_Free(S_new);
}

File: R/survAUC_HungChiang.R
################################################################
###						Hung & Chiang						####
################################################################
## Surv.rsp		- the Surv(.,.) Outcome of training data
## Surv.rsp.new	- the Surv(.,.) Outcome of test data
## lpnew		- the vector of linear predictors of test data
## times		- the vector of times



#' @title AUC estimator proposed by Hung and Chiang
#' 
#' @rdname survAUC_HungChiang
#' @name survAUC_HungChiang
#' 
#' @description Hung and Chiang's estimator of cumulative/dynamic AUC for right-censored
#' time-to-event data
#' 
#' @details This function implements the estimator of cumulative/dynamic AUC proposed by
#' Hung and Chiang (2010). The estimator is based on
#' inverse-probability-of-censoring weights and does not assume a specific
#' working model for deriving the predictor \code{lpnew}. It is assumed,
#' however, that there is a one-to-one relationship between the predictor and
#' the expected survival times conditional on the predictor. The \code{iauc}
#' summary measure is given by the integral of AUC on [0, max(\code{times})]
#' (weighted by the estimated probability density of the time-to-event
#' outcome).
#' 
#' Note that the estimator implemented in \code{AUC.hc} is restricted to
#' situations where the random censoring assumption holds (formula (4) in Hung
#' and Chiang 2010).
#' 
#' @param Surv.rsp A \code{Surv(.,.)} object containing to the outcome of the
#' training data.
#' @param Surv.rsp.new A \code{Surv(.,.)} object containing the outcome of the
#' test data.
#' @param lpnew The vector of predictors obtained from the test data.
#' @param times A vector of time points at which to evaluate AUC.
#' @return \code{AUC.hc} returns an object of class \code{survAUC}.
#' Specifically, \code{AUC.hc} returns a list with the following components:
#' \item{auc}{The cumulative/dynamic AUC estimates (evaluated at
#' \code{times}).} \item{times}{The vector of time points at which AUC is
#' evaluated.} \item{iauc}{The summary measure of AUC.}
#' @seealso \code{\link{AUC.uno}}, \code{\link{AUC.sh}}, \code{\link{AUC.cd}},
#' \code{\link{IntAUC}}
#' @references
#' 
#' Hung, H. and C.-T. Chiang (2010). \cr Estimation methods for time-dependent
#' AUC models with survival data.\cr \emph{Canadian Journal of Statistics}
#' \bold{38}, 8--26.\cr
#' @keywords classif
#' @examples
#' 
#' data(cancer,package="survival")
#' TR <- ovarian[1:16,]
#' TE <- ovarian[17:26,]
#' train.fit  <- survival::coxph(survival::Surv(futime, fustat) ~ age,
#'                     x=TRUE, y=TRUE, method="breslow", data=TR)
#' 
#' lpnew <- predict(train.fit, newdata=TE)
#' Surv.rsp <- survival::Surv(TR$futime, TR$fustat)
#' Surv.rsp.new <- survival::Surv(TE$futime, TE$fustat)
#' times <- seq(10, 1000, 10)                  
#' 
#' AUC_hc <- AUC.hc(Surv.rsp, Surv.rsp.new, lpnew, times)
#' AUC_hc
#' 
#' 
#' @export AUC.hc
AUC.hc <- function(Surv.rsp, Surv.rsp.new, lpnew, times)
{
## Surv-train
	stime <- Surv.rsp[,1]
	event <- Surv.rsp[,2]
	
## Surv-test
	stime.new <- Surv.rsp.new[,1]
	event.new <- Surv.rsp.new[,2]

	n_time <- length(times)
	n_stime <- length(stime)
	n_stime_new <- length(stime.new)
	n_lpnew <- length(lpnew)
	auc <- vector("numeric",n_time)
	
	ans <- .C(`C_Hung_Chiang`,
			  as.numeric(times),
			  as.integer(n_time),
			  as.numeric(stime),
			  as.numeric(event),
			  as.integer(n_stime),
			  as.numeric(stime.new),
			  as.numeric(event.new),
			  as.integer(n_stime_new),
			  as.numeric(lpnew),
			  as.integer(n_lpnew),
			  as.numeric(auc),
			  as.numeric(0))
	#No longer needed since the symbol is registered in the NAMESPACE
	#          ,PACKAGE="survAUC")
	erg <- list(auc=ans[[11]], times=times, iauc=ans[[12]])
	class(erg) <- "survAUC"
	erg
}


